File: src/main/java/org/trp/shincolle/server/TeamDiplomacyService.java
package org.trp.shincolle.server;

import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.server.players.GameProfileCache;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.neoforged.neoforge.network.PacketDistributor;
import org.trp.shincolle.attachment.AdmiralData;
import org.trp.shincolle.entity.base.EntityShipBase;
import org.trp.shincolle.network.C2STeamDiplomacyPayload;
import org.trp.shincolle.network.S2CDeskDiplomacySyncPayload;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.UUID;

public final class TeamDiplomacyService {
    private TeamDiplomacyService() {
    }

    public static boolean isDiplomaticAlly(EntityShipBase ship, Entity target) {
        if (ship == null || target == null) {
            return false;
        }
        if (!(ship.level() instanceof ServerLevel serverLevel)) {
            return false;
        }
        UUID owner = ship.getOwnerUUID();
        UUID targetOwner = PointerInteractionService.getTargetOwnerUUID(target);
        return TeamDiplomacySavedData.get(serverLevel).areAllies(owner, targetOwner);
    }

    public static boolean isDiplomaticBanned(EntityShipBase ship, Entity target) {
        if (ship == null || target == null) {
            return false;
        }
        if (!(ship.level() instanceof ServerLevel serverLevel)) {
            return false;
        }
        UUID owner = ship.getOwnerUUID();
        UUID targetOwner = PointerInteractionService.getTargetOwnerUUID(target);
        return TeamDiplomacySavedData.get(serverLevel).isBanned(owner, targetOwner);
    }

    public static void handleAction(Player player, int action, UUID target) {
        if (player == null) {
            return;
        }
        if (!(player.level() instanceof ServerLevel serverLevel)) {
            return;
        }

        UUID owner = player.getUUID();
        if (target == null || owner.equals(target)) {
            return;
        }

        TeamDiplomacySavedData diplomacy = TeamDiplomacySavedData.get(serverLevel);
        Player targetPlayer = serverLevel.getPlayerByUUID(target);
        Component targetName = targetPlayer != null ? targetPlayer.getDisplayName() : Component.literal(target.toString());
        boolean changed;
        Component message;
        switch (action) {
            case C2STeamDiplomacyPayload.ACTION_ADD_ALLY -> {
                changed = diplomacy.addAlly(owner, target);
                message = changed
                        ? Component.translatable("chat.shincolle.team.ally_added").append(targetName)
                        : Component.translatable("chat.shincolle.team.ally_unchanged").append(targetName);
            }
            case C2STeamDiplomacyPayload.ACTION_REMOVE_ALLY -> {
                changed = diplomacy.removeAlly(owner, target);
                message = changed
                        ? Component.translatable("chat.shincolle.team.ally_removed").append(targetName)
                        : Component.translatable("chat.shincolle.team.ally_missing").append(targetName);
            }
            case C2STeamDiplomacyPayload.ACTION_ADD_BANNED -> {
                changed = diplomacy.addBanned(owner, target);
                message = changed
                        ? Component.translatable("chat.shincolle.team.hostile_added").append(targetName)
                        : Component.translatable("chat.shincolle.team.hostile_unchanged").append(targetName);
            }
            case C2STeamDiplomacyPayload.ACTION_REMOVE_BANNED -> {
                changed = diplomacy.removeBanned(owner, target);
                message = changed
                        ? Component.translatable("chat.shincolle.team.hostile_removed").append(targetName)
                        : Component.translatable("chat.shincolle.team.hostile_missing").append(targetName);
            }
            default -> {
                return;
            }
        }

        player.displayClientMessage(message, false);
        if (changed && player instanceof ServerPlayer serverPlayer) {
            sendDeskDiplomacySync(serverPlayer);
        }
    }

    public static void sendDeskDiplomacySync(ServerPlayer player) {
        TeamDiplomacySavedData diplomacy = TeamDiplomacySavedData.get(player.serverLevel());
        updateDiplomacyDisplayData(player, diplomacy);
        TeamDiplomacySavedData.TeamDiplomacyEntry entry = diplomacy.getOrCreate(player.getUUID());

        LinkedHashSet<UUID> displayIds = new LinkedHashSet<>();
        displayIds.addAll(entry.allies());
        displayIds.addAll(entry.banned());

        ArrayList<UUID> uuids = new ArrayList<>();
        ArrayList<String> teamNames = new ArrayList<>();
        ArrayList<String> leaderNames = new ArrayList<>();
        for (UUID target : displayIds) {
            if (target == null) {
                continue;
            }
            uuids.add(target);
            TeamDiplomacySavedData.TeamDiplomacyEntry targetEntry = diplomacy.get(target);
            teamNames.add(targetEntry == null ? "" : targetEntry.teamName());
            String leaderName = targetEntry == null ? "" : targetEntry.leaderName();
            if (leaderName.isBlank()) {
                leaderName = resolveDiplomacyLeaderName(player, target);
            }
            leaderNames.add(leaderName);
        }

        PacketDistributor.sendToPlayer(player, S2CDeskDiplomacySyncPayload.of(
                player.getUUID(),
                entry.allies(),
                entry.banned(),
                uuids,
                teamNames,
                leaderNames
        ));
    }

    private static void updateDiplomacyDisplayData(ServerPlayer player, TeamDiplomacySavedData diplomacy) {
        AdmiralData data = PlayerStateService.admiralData(player);
        String teamName = data.getTeamName(data.getCurrentTeamID());
        String leaderName = player.getName().getString();
        diplomacy.setDisplayData(player.getUUID(), teamName, leaderName);
    }

    private static String resolveDiplomacyLeaderName(ServerPlayer player, UUID target) {
        if (target == null) {
            return "";
        }
        ServerPlayer onlinePlayer = player.server.getPlayerList().getPlayer(target);
        if (onlinePlayer != null) {
            return onlinePlayer.getName().getString();
        }
        GameProfileCache profileCache = player.server.getProfileCache();
        if (profileCache == null) {
            return "";
        }
        return profileCache.get(target).map(com.mojang.authlib.GameProfile::getName).orElse("");
    }
}

File: src/main/java/org/trp/shincolle/server/TeamDiplomacyService.kt
package org.trp.shincolle.server

import com.mojang.authlib.GameProfile
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.neoforged.neoforge.network.PacketDistributor
import org.trp.shincolle.entity.base.EntityShipBase
import org.trp.shincolle.network.C2STeamDiplomacyPayload
import org.trp.shincolle.network.S2CDeskDiplomacySyncPayload
import java.util.*

object TeamDiplomacyService {
    @JvmStatic
    fun isDiplomaticAlly(ship: EntityShipBase?, target: Entity?): Boolean {
        if (ship == null || target == null) {
            return false
        }
        if (ship.level() !is ServerLevel) {
            return false
        }
        val serverLevel = ship.level() as ServerLevel
        val owner = ship.ownerUUID
        val targetOwner = PointerInteractionService.getTargetOwnerUUID(target)
        return TeamDiplomacySavedData.Companion.get(serverLevel).areAllies(owner, targetOwner)
    }

    @JvmStatic
    fun isDiplomaticBanned(ship: EntityShipBase?, target: Entity?): Boolean {
        if (ship == null || target == null) {
            return false
        }
        if (ship.level() !is ServerLevel) {
            return false
        }
        val serverLevel = ship.level() as ServerLevel
        val owner = ship.ownerUUID
        val targetOwner = PointerInteractionService.getTargetOwnerUUID(target)
        return TeamDiplomacySavedData.Companion.get(serverLevel).isBanned(owner, targetOwner)
    }

    @JvmStatic
    fun handleAction(player: Player?, action: Int, target: UUID?) {
        if (player == null) {
            return
        }
        if (player.level() !is ServerLevel) {
            return
        }

        val serverLevel = player.level() as ServerLevel
        val owner = player.uuid
        if (target == null || owner == target) {
            return
        }

        val diplomacy: TeamDiplomacySavedData = TeamDiplomacySavedData.Companion.get(serverLevel)
        val targetPlayer: Player? = serverLevel.getPlayerByUUID(target)
        val targetName: Component =
            (targetPlayer?.displayName ?: net.minecraft.network.chat.Component.literal(target.toString()))
        val changed = applyDiplomacyAction(diplomacy, owner, action, target)
        val message: Component?
        when (action) {
            C2STeamDiplomacyPayload.ACTION_ADD_ALLY -> {
                message = if (changed)
                    Component.translatable("chat.shincolle.team.ally_added").append(targetName)
                else
                    Component.translatable("chat.shincolle.team.ally_unchanged").append(targetName)
            }

            C2STeamDiplomacyPayload.ACTION_REMOVE_ALLY -> {
                message = if (changed)
                    Component.translatable("chat.shincolle.team.ally_removed").append(targetName)
                else
                    Component.translatable("chat.shincolle.team.ally_missing").append(targetName)
            }

            C2STeamDiplomacyPayload.ACTION_ADD_BANNED -> {
                message = if (changed)
                    Component.translatable("chat.shincolle.team.hostile_added").append(targetName)
                else
                    Component.translatable("chat.shincolle.team.hostile_unchanged").append(targetName)
            }

            C2STeamDiplomacyPayload.ACTION_REMOVE_BANNED -> {
                message = if (changed)
                    Component.translatable("chat.shincolle.team.hostile_removed").append(targetName)
                else
                    Component.translatable("chat.shincolle.team.hostile_missing").append(targetName)
            }

            else -> {
                return
            }
        }

        player.displayClientMessage(message, false)
        if (changed && player is ServerPlayer) {
            sendDeskDiplomacySync(player)
        }
    }

    fun applyDiplomacyAction(diplomacy: TeamDiplomacySavedData, owner: UUID?, action: Int, target: UUID?): Boolean {
        return when (action) {
            C2STeamDiplomacyPayload.ACTION_ADD_ALLY -> diplomacy.addAlly(owner, target)
            C2STeamDiplomacyPayload.ACTION_REMOVE_ALLY -> diplomacy.removeAlly(owner, target)
            C2STeamDiplomacyPayload.ACTION_ADD_BANNED -> diplomacy.addBanned(owner, target)
            C2STeamDiplomacyPayload.ACTION_REMOVE_BANNED -> diplomacy.removeBanned(owner, target)
            else -> false
        }
    }

    @JvmStatic
    fun sendDeskDiplomacySync(player: ServerPlayer) {
        val diplomacy: TeamDiplomacySavedData = TeamDiplomacySavedData.Companion.get(player.serverLevel())
        updateDiplomacyDisplayData(player, diplomacy)
        val entry = diplomacy.getOrCreate(player.uuid)

        val displayIds = LinkedHashSet<UUID?>()
        displayIds.addAll(entry.allies)
        displayIds.addAll(entry.banned)

        val uuids = ArrayList<UUID?>()
        val teamNames = ArrayList<String?>()
        val leaderNames = ArrayList<String?>()
        for (target in displayIds) {
            if (target == null) {
                continue
            }
            uuids.add(target)
            val targetEntry = diplomacy.get(target)
            teamNames.add(targetEntry?.teamName ?: "")
            var leaderName = targetEntry?.leaderName ?: ""
            if (leaderName.isBlank()) {
                leaderName = resolveDiplomacyLeaderName(player, target)
            }
            leaderNames.add(leaderName)
        }

        PacketDistributor.sendToPlayer(
            player, S2CDeskDiplomacySyncPayload.of(
                player.uuid,
                ArrayList(entry.allies),
                ArrayList(entry.banned),
                uuids,
                teamNames,
                leaderNames
            )
        )
    }

    private fun updateDiplomacyDisplayData(player: ServerPlayer, diplomacy: TeamDiplomacySavedData) {
        val data = PlayerStateService.admiralData(player)
        val teamName = data.getTeamName(data.getCurrentTeamID())
        val leaderName = player.name.string
        diplomacy.setDisplayData(player.uuid, teamName, leaderName)
    }

    private fun resolveDiplomacyLeaderName(player: ServerPlayer, target: UUID?): String {
        if (target == null) {
            return ""
        }
        val onlinePlayer = player.server.playerList.getPlayer(target)
        if (onlinePlayer != null) {
            return onlinePlayer.name.string
        }
        val profileCache = player.server.profileCache
        if (profileCache == null) {
            return ""
        }
        return profileCache.get(target).map { it.name }.orElse("")
    }
}
